// sandstone/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagClass = exports.ResourceClass = exports.RecipeClass = exports.PredicateClass = exports.LootTableClass = exports.AdvancementClass = exports.BasePathClass = void 0;
__exportStar(require("./_internals/arguments/index"), exports);
var BasePath_1 = require("./_internals/datapack/BasePath");
Object.defineProperty(exports, "BasePathClass", { enumerable: true, get: function () { return BasePath_1.BasePathClass; } });
var index_1 = require("./_internals/resources/index");
Object.defineProperty(exports, "AdvancementClass", { enumerable: true, get: function () { return index_1.AdvancementClass; } });
Object.defineProperty(exports, "LootTableClass", { enumerable: true, get: function () { return index_1.LootTableClass; } });
Object.defineProperty(exports, "PredicateClass", { enumerable: true, get: function () { return index_1.PredicateClass; } });
Object.defineProperty(exports, "RecipeClass", { enumerable: true, get: function () { return index_1.RecipeClass; } });
Object.defineProperty(exports, "ResourceClass", { enumerable: true, get: function () { return index_1.ResourceClass; } });
Object.defineProperty(exports, "TagClass", { enumerable: true, get: function () { return index_1.TagClass; } });
