// sandstone/variables.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Variable = exports.Selector = exports.createObjective = exports.SelectorClass = exports.rel = exports.loc = exports.abs = void 0;
const index_1 = require("./_internals/index");
var index_2 = require("./_internals/index");
Object.defineProperty(exports, "abs", { enumerable: true, get: function () { return index_2.absolute; } });
Object.defineProperty(exports, "loc", { enumerable: true, get: function () { return index_2.local; } });
Object.defineProperty(exports, "rel", { enumerable: true, get: function () { return index_2.relative; } });
__exportStar(require("./_internals/variables/Coordinates"), exports);
__exportStar(require("./_internals/variables/JsonTextComponentClass"), exports);
__exportStar(require("./_internals/variables/NBTs"), exports);
__exportStar(require("./_internals/variables/Objective"), exports);
__exportStar(require("./_internals/variables/parsers"), exports);
__exportStar(require("./_internals/variables/PlayerScore"), exports);
var Selector_1 = require("./_internals/variables/Selector");
Object.defineProperty(exports, "SelectorClass", { enumerable: true, get: function () { return Selector_1.SelectorClass; } });
exports.createObjective = index_1.dataPack.createObjective, exports.Selector = index_1.dataPack.Selector, exports.Variable = index_1.dataPack.Variable;
